File: 12319 Edgetown's Traffic Jams/solution.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#define INF 100
#define MAXV 100

int **dist, **newDist;
char *buffer;
size_t bufsize;

void input(int **array, int V) {
    for(int u = 1; u <= V; ++u) {
        scanf("%*d");
        getline(&buffer, &bufsize, stdin);
        char *token = strtok(buffer, " ");
        while(token) {
            int v = atoi(token);
            array[u][v] = 1;
            token = strtok(NULL, " ");
        }
    }
}

void floydWarshall(int **array, int V) {
    for(int k = 1; k <= V; ++k) {
        for(int i = 1; i <= V; ++i) {
            for(int j = 1; j <= V; ++j) {
                if(array[i][j] > array[i][k] + array[k][j])
                    array[i][j] = array[i][k] + array[k][j];
            }
        }
    }
}

bool accept(int **dist, int **newDist, int V, int A, int B) {
    for(int i = 1; i <= V; ++i) {
        for(int j = 1; j <= V; ++j) {
            if(newDist[i][j] > A * dist[i][j] + B)
                return false;
        }
    }
    return true;
}

int main() {
    int **dist = (int**)malloc((MAXV + 1) * sizeof(int*)), **newDist = (int**)malloc((MAXV + 1) * sizeof(int*));
    for(int i = 1; i <= MAXV; ++i) {
        dist[i] = (int*)malloc((MAXV + 1) * sizeof(int));
        newDist[i] = (int*)malloc((MAXV + 1) * sizeof(int));
    }
    int n;
    while(scanf("%d", &n) && n != 0) {
        for(int i = 1; i <= n; ++i) {
            for(int j = 1; j <= n; ++j)
                dist[i][j] = newDist[i][j] = i == j ? 0 : INF;
        }
        input(dist, n);
        floydWarshall(dist, n);
        input(newDist, n);
        floydWarshall(newDist, n);
        int A, B;
        scanf("%d %d", &A, &B);
        puts(accept(dist, newDist, n, A, B) ? "Yes" : "No");
    }
    free(buffer);
    for(int i = 1; i <= MAXV; ++i) {
        free(dist[i]);
        free(newDist[i]);
    }
    free(dist);
    free(newDist);
	return 0;
}

File: 12319 Edgetown's Traffic Jams/solution.java
import java.io.*;

public class Main {
    static int INF = 100, MAXV = 100;
    static int[][] dist = new int[MAXV + 1][MAXV + 1], newDist = new int[MAXV + 1][MAXV + 1];
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static StreamTokenizer st = new StreamTokenizer(br);
	public static void main(String[] args) throws IOException {
		int n;
        StringBuilder output = new StringBuilder();
        while(st.nextToken() == StreamTokenizer.TT_NUMBER && (n = (int)st.nval) != 0) {
            for(int i = 1; i <= n; ++i) {
                for(int j = 1; j <= n; ++j)
                    dist[i][j] = newDist[i][j] = i == j ? 0 : INF;
            }
            input(dist, n);
            floydWarshall(dist, n);
            input(newDist, n);
            floydWarshall(newDist, n);
            st.nextToken();
            int A = (int)st.nval;
            st.nextToken();
            int B = (int)st.nval;
            output.append(accept(n, A, B) ? "Yes\n" : "No\n");
        }
	    System.out.print(output);
	}
	
	static void input(int[][] array, int V) throws IOException {
	    for(int u = 1; u <= V; ++u) {
	        st.nextToken();
	        String[] tokens = br.readLine().split("\\s+");
	        for(String s : tokens)
	            array[u][Integer.parseInt(s)] = 1;
	    }
	}
	
	static void floydWarshall(int[][] array, int V) {
        for(int k = 1; k <= V; ++k) {
            for(int i = 1; i <= V; ++i) {
                for(int j = 1; j <= V; ++j)
                    array[i][j] = Math.min(array[i][j], array[i][k] + array[k][j]);
            }
        }
    }
    
    static boolean accept(int V, int A, int B) {
        for(int i = 1; i <= V; ++i) {
            for(int j = 1; j <= V; ++j) {
                if(newDist[i][j] > A * dist[i][j] + B)
                    return false;
            }
        }
        return true;
    }
}
